File: packages/flutter/lib/src/mlkit_detector.dart
import 'package:google_mlkit_face_detection/google_mlkit_face_detection.dart';
import 'types.dart';

/// ML Kit face detector with landmark extraction.
///
/// Converts ML Kit [Face] objects into the SDK's [Detection] type.
/// Keypoint mapping:
///   FaceLandmarkType.leftEye  → [1] (camera-left = subject's right eye → index 0)
///   FaceLandmarkType.rightEye → [0]
///   FaceLandmarkType.noseBase → [2]
class MLKitFaceDetector {
  MLKitFaceDetector()
      : _detector = FaceDetector(
          options: FaceDetectorOptions(
            enableLandmarks:    true,
            enableClassification: false,
            enableTracking:     true,
            minFaceSize:        0.1,
            performanceMode:    FaceDetectorMode.accurate,
          ),
        );

  final FaceDetector _detector;

  Future<List<Detection>> detectFromInputImage(
    InputImage image,
    int frameWidth,
    int frameHeight,
  ) async {
    final faces = await _detector.processImage(image);
    return faces.map((face) => _convert(face, frameWidth, frameHeight)).toList();
  }

  Detection _convert(Face face, int w, int h) {
    final r = face.boundingBox;

    // Normalise bounding box
    final originX = r.left   / w;
    final originY = r.top    / h;
    final bWidth  = r.width  / w;
    final bHeight = r.height / h;

    final cx = originX + bWidth  / 2;
    final cy = originY + bHeight / 2;

    // Landmarks — fall back to estimated positions if not detected
    Keypoint kp(FaceLandmarkType type, double defaultX, double defaultY) {
      final lm = face.landmarks[type];
      if (lm == null) return Keypoint(x: defaultX, y: defaultY);
      return Keypoint(x: lm.position.x / w, y: lm.position.y / h);
    }

    final rightEye = kp(FaceLandmarkType.rightEye, cx - bWidth * 0.15, cy - bHeight * 0.1);
    final leftEye  = kp(FaceLandmarkType.leftEye,  cx + bWidth * 0.15, cy - bHeight * 0.1);
    final noseTip  = kp(FaceLandmarkType.noseBase, cx,                 cy + bHeight * 0.05);

    return Detection(
      boundingBox: BoundingBox(
        originX: originX,
        originY: originY,
        width:   bWidth,
        height:  bHeight,
      ),
      keypoints: [rightEye, leftEye, noseTip],
      score:     1.0,  // ML Kit doesn't expose confidence directly
    );
  }

  void dispose() => _detector.close();
}

File: packages/react-native/src/detectors/mlkit-detector.ts
import type { Detection } from "@meeeetup1120/core";

/**
 * ML Kit face detection adapter for React Native.
 *
 * Install peer deps:
 *   npx expo install @react-native-ml-kit/face-detection
 *   # or for bare RN:
 *   npm i @react-native-ml-kit/face-detection
 *
 * ML Kit keypoint labels → our index mapping:
 *   LEFT_EYE  → [1] (camera-left = subject's left = rightEye in mirror)
 *   RIGHT_EYE → [0]
 *   NOSE_BASE → [2]
 *
 * We normalise all coordinates to 0–1 using the frame dimensions.
 */

interface MLKitFace {
  bounds: { left: number; top: number; width: number; height: number };
  landmarks?: Record<string, { position: { x: number; y: number } }>;
  headEulerAngleY?: number;
  leftEyeOpenProbability?: number;
  rightEyeOpenProbability?: number;
  trackingId?: number;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function mlkit(): any {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  return require("@react-native-ml-kit/face-detection");
}

export async function detectFacesMLKit(
  imagePath: string,
  frameWidth: number,
  frameHeight: number,
): Promise<Detection[]> {
  const FaceDetector = mlkit().default;
  const faces: MLKitFace[] = await FaceDetector.detect(imagePath);

  return faces.map((face) => {
    const b = face.bounds;
    const cx = (b.left + b.width  / 2) / frameWidth;
    const cy = (b.top  + b.height / 2) / frameHeight;
    const bw = b.width  / frameWidth;
    const bh = b.height / frameHeight;

    const lm = face.landmarks ?? {};
    const kp = (label: string) => {
      const pos = lm[label]?.position;
      return pos ? { x: pos.x / frameWidth, y: pos.y / frameHeight } : null;
    };

    const rightEye = kp("RIGHT_EYE") ?? { x: cx - bw * 0.15, y: cy - bh * 0.1 };
    const leftEye  = kp("LEFT_EYE")  ?? { x: cx + bw * 0.15, y: cy - bh * 0.1 };
    const noseTip  = kp("NOSE_BASE") ?? { x: cx,              y: cy + bh * 0.05 };

    return {
      boundingBox: {
        originX: b.left / frameWidth,
        originY: b.top  / frameHeight,
        width:   bw,
        height:  bh,
      },
      keypoints: [rightEye, leftEye, noseTip],
      score: 1,  // ML Kit doesn't expose confidence directly
    };
  });
}
